### DISCUSSION TITLE
PNP Search Returning Spurious Results

### DISCUSSION THREAD

**Discussion by StephenHoughton4** [microsoft-search/pnp-modern-search#2922] [General]
I have a PNP Search now integrated into my SharePoint, so we can add a Personalised People Search. The main addition we have added is the Displayed Name will show the User Account Initials (e.g. **John Smith (JTS)** ). 

From reading previous Discussions, I can see the search box text cannot be separated into Word1 OR Word2, without writing the word OR in the search term. (Ref: [https://github.com/microsoft-search/pnp-modern-search/discussions/2909](#2909 ))

So With that, I need some assistance on how to proceed, to make the search act as intended. If we leave the standard set up and we search FAL DUBLIN, we get one user (Jimmy Fallon from the Dublin Office), but if we Search FAL we get a list of users (some Office: Falklands, JobTitle: Falconer, Last Name Fal... etc), however we also get users that do not seem to have any Properties starting with FAL (E.g. Liz Taylor, Office: Belgium, JobTitle: Actor). We dont want these seemingly random people in the results.

Please could you assist?

Our Result Source has the following Query Modification Text:
> ContentClass=urn:content-class:SPSPeople (AADObjectID:a* OR AADObjectID:b* OR AADObjectID:c* OR AADObjectID:d* OR AADObjectID:e* OR AADObjectID:f* OR AADObjectID:g* OR AADObjectID:h* OR AADObjectID:i* OR AADObjectID:j* OR AADObjectID:k* OR AADObjectID:l* OR AADObjectID:m* OR AADObjectID:n* OR AADObjectID:o* OR AADObjectID:p* OR AADObjectID:q* OR AADObjectID:r* OR AADObjectID:s* OR AADObjectID:t* OR AADObjectID:u* OR AADObjectID:v* OR AADObjectID:w* OR AADObjectID:x* OR AADObjectID:y* OR AADObjectID:z* OR AADObjectID:1* OR AADObjectID:2* OR AADObjectID:3* OR AADObjectID:4* OR AADObjectID:5* OR AADObjectID:6* OR AADObjectID:7* OR AADObjectID:8* OR AADObjectID:9* OR AADObjectID:0*)
    
Our Search Results - Query Template is as follows:

> ({searchTerms}*) OR ({searchTerms}) 

This returns the randomly named people when searching for FAL, but returns the 1 match when searching FAL Dublin.

I attempted to amend the Search Results - Query Template as follows:
> {|AccountName:(	{searchTerms}*) OR Office:({searchTerms}*)

Which worked When Searching FAL to bring back people name FAL* (or Office: Falklands). However if I wanted to narrow the search to FAL DUBLIN, this would not return anything. I worked on this angle for a while thinking it was the correct method, but due to the Query not being editable (to make it FAL or DUBLIN) I need assistance on what to try next.

I assume then my next attempt would be to edit the Search Query modification to only call certain columns, rather than searching all Properties. Could you suggest what the Result Source - Query Modification text should be, to only search Firstname / LastName and Office location (I will extend these afterwards to include job title / phone number etc but wanted a start point).

Thanks
Stephen

**Comment by StephenHoughton4** [comment]
Search Results Set-Up:

Selected Properties
AuthorOWSUSER, CompanyOWSTEXT, contentclass, ContentTypeId, Created,CreatedBy, DefaultEncodingURL, DocumentLink, FileType, JobTitle, NormListId, NormSiteId, NormUniqueId, NormWebId, OrgNames, owstaxidmetadataalltagsifo, Path, Site Title, SPSiteUrl, Title, UniqueId, UserName, WorkPhone
I also Added HitHighlightedSummary when debugging, to see what was being hit when searching (but didn't seem to help).

Sort Settings:
![image](https://user-images.githubusercontent.com/129411416/233679919-fca4a157-199b-4bc9-9887-7716fae37f97.png)

Enable Lazy Searching: On

**Reply by wobba** [reply]
What's the point of the `AADObjectID` parts? And seems you want either of the terms to match in one or more fields, is that correct? And I would use the LocalPeople result source and put everything else in the web part query template.

And I have to think some on how to make this work for you (if possible) without a custom query extension which would work..as that could rewrite the input any way you want.

**Reply by StephenHoughton4** [reply]
Hi @wobba 
It is correct that we are searching 2 properties at the same time (Location and LastName), but we also would search for just LastName as well. The AADObjectID part of the query I am not sure, I thought that was standard, are you saying to try removing that?

Any help would be appreciated. Thanks :)

**Reply by StephenHoughton4** [reply]
Hi @wobba 

I have found the cause of the spurious results. Search is searching "Memberships", so any AD Groups they are part of. Is there a way to exclude Membership Groups from the Search? Or is it a case of modifying the import from AD, to exclude the membership groups?

Either way help would be appreciated :) Thanks

**Reply by StephenHoughton4** [reply]
![image](https://user-images.githubusercontent.com/129411416/234340868-7af4c6ea-c5ce-4aed-84d6-dec7f5b8f6d7.png)
Using this Search Query produces the results I would Like, but I do have a concern, 

Example:
If I were to search: "DEAN"
Dean Davies Who is in Membership "DeanOfSciences" may not show in the results, so not sure this is the best resolution. Hoping someone can provide assistance (feel I am close, but not there).

Thanks.

**Comment by kasperbolarsen** [comment]
I have made an Employee directory for more than 50 customers in the last few years,  and one of most commonly used "tricks" in a case like yours has been to search in all fields but boost the ranking usung XRANK when the query matches specific properties. 

Bing Search for xrank or find inspiration here https://www.techmikael.com/2016/12/tip-for-writing-multiple-xranks-or.html?m=1

**Reply by StephenHoughton4** [reply]
Thanks for the suggestion @kasperbolarsen . I couldn't seem to work out where to enter this XRANK formula.

If you can advise, I will give it a go :) 

Thanks

**Reply by kasperbolarsen** [reply]
Talking about good timing 🙂 this post was released today : https://www.techmikael.com/2023/04/retirement-of-dynamic-ordering-feature.html

**Comment by StephenHoughton4** [comment]
Thanks to a few different suggestions and testing methods I ended up doing the following:

`({SearchTerms}* AND NOT Memberships:{SearchTerms}*)
OR UserName:({searchTerms}*) 
OR JobTitle:({searchTerms}*) 
OR AccountName:({searchTerms}*) 
OR Department:({searchTerms}*) 
OR Office:({searchTerms}*) 
OR Title:({searchTerms}*) 
OR PersonInitials:({searchTerms}*)
`

I did not know of the NOT clause, but then adding this with the Or statements meant I was so close to the answer.. This seems to have done the trick.

Thanks @wobba  and @kasperbolarsen  for the suggestions, I will be integrating the XRANK feature as well, but this is the core Code I used.

As a side note I added

`LastName:*` 

To Remove any entries where they had Null values in the Last Name field (Mailbox Accounts).

**Reply by wobba** [reply]
You should always have () around any `{SearchTerms}*` to ensure correct boolean parsing when using multi terms. And propose to add () around all logical parts so it's evaluated the way you intend always. () are cheap, but makes it clearer to read as well. The first part should be:

`(({SearchTerms}*) AND NOT Memberships:({SearchTerms}*))`
